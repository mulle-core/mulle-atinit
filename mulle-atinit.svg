<?xml version="1.0" encoding="UTF-8"?><svg xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" fill-opacity="1" color-rendering="auto" color-interpolation="auto" text-rendering="auto" stroke="black" stroke-linecap="square" width="756" stroke-miterlimit="10" shape-rendering="auto" stroke-opacity="1" fill="black" stroke-dasharray="none" font-weight="normal" stroke-width="1" height="273" font-family="'Dialog'" font-style="normal" stroke-linejoin="miter" font-size="12px" stroke-dashoffset="0" image-rendering="auto">
  <!--Generated by ySVG 2.5-->
  <defs id="genericDefs"/>
  <g>
    <defs id="defs1">
      <clipPath clipPathUnits="userSpaceOnUse" id="clipPath1">
        <path d="M0 0 L756 0 L756 273 L0 273 L0 0 Z"/>
      </clipPath>
      <clipPath clipPathUnits="userSpaceOnUse" id="clipPath2">
        <path d="M-20 186 L736 186 L736 459 L-20 459 L-20 186 Z"/>
      </clipPath>
    </defs>
    <g fill="white" text-rendering="geometricPrecision" shape-rendering="geometricPrecision" transform="translate(20,-186)" stroke="white">
      <rect x="-20" width="756" height="273" y="186" clip-path="url(#clipPath2)" stroke="none"/>
    </g>
    <g fill="rgb(204,255,255)" text-rendering="geometricPrecision" shape-rendering="geometricPrecision" transform="matrix(1,0,0,1,20,-186)" stroke="rgb(204,255,255)">
      <rect x="-5" width="158" height="183" y="231" clip-path="url(#clipPath2)" stroke="none"/>
    </g>
    <g text-rendering="geometricPrecision" stroke-miterlimit="1.45" shape-rendering="geometricPrecision" transform="matrix(1,0,0,1,20,-186)" stroke-linecap="butt">
      <rect fill="none" x="-5" width="158" height="183" y="231" clip-path="url(#clipPath2)"/>
      <text x="5.5994" xml:space="preserve" y="294.4678" clip-path="url(#clipPath2)" font-family="'Noto Sans Mono'" stroke="none">static void  foo()</text>
      <text x="5.5994" xml:space="preserve" y="310.8119" clip-path="url(#clipPath2)" font-family="'Noto Sans Mono'" stroke="none">{</text>
      <text x="5.5994" xml:space="preserve" y="327.156" clip-path="url(#clipPath2)" font-family="'Noto Sans Mono'" stroke="none">}</text>
      <text x="5.5994" xml:space="preserve" y="343.5002" clip-path="url(#clipPath2)" font-family="'Noto Sans Mono'" stroke="none"> </text>
      <text x="5.5994" xml:space="preserve" y="359.8443" clip-path="url(#clipPath2)" font-family="'Noto Sans Mono'" stroke="none">mulle_atinit( foo);</text>
    </g>
    <g fill="rgb(255,204,0)" text-rendering="geometricPrecision" shape-rendering="geometricPrecision" transform="matrix(1,0,0,1,20,-186)" stroke="rgb(255,204,0)">
      <rect x="-5" width="158" height="30" y="414" clip-path="url(#clipPath2)" stroke="none"/>
    </g>
    <g text-rendering="geometricPrecision" stroke-miterlimit="1.45" shape-rendering="geometricPrecision" transform="matrix(1,0,0,1,20,-186)" stroke-linecap="butt">
      <rect fill="none" x="-5" width="158" height="30" y="414" clip-path="url(#clipPath2)"/>
      <text x="59.9287" xml:space="preserve" y="433.1543" clip-path="url(#clipPath2)" font-family="sans-serif" stroke="none">shlib</text>
    </g>
    <g fill="rgb(204,255,255)" text-rendering="geometricPrecision" shape-rendering="geometricPrecision" transform="matrix(1,0,0,1,20,-186)" stroke="rgb(204,255,255)">
      <rect x="200.5" width="211" height="183" y="231" clip-path="url(#clipPath2)" stroke="none"/>
    </g>
    <g text-rendering="geometricPrecision" stroke-miterlimit="1.45" shape-rendering="geometricPrecision" transform="matrix(1,0,0,1,20,-186)" stroke-linecap="butt">
      <rect fill="none" x="200.5" width="211" height="183" y="231" clip-path="url(#clipPath2)"/>
      <text x="255.5996" xml:space="preserve" y="302.6399" clip-path="url(#clipPath2)" font-family="'Noto Sans Mono'" stroke="none">int   main()</text>
      <text x="255.5996" xml:space="preserve" y="318.984" clip-path="url(#clipPath2)" font-family="'Noto Sans Mono'" stroke="none">{</text>
      <text x="255.5996" xml:space="preserve" y="335.3281" clip-path="url(#clipPath2)" font-family="'Noto Sans Mono'" stroke="none">   return( 0);</text>
      <text x="255.5996" xml:space="preserve" y="351.6722" clip-path="url(#clipPath2)" font-family="'Noto Sans Mono'" stroke="none">}</text>
    </g>
    <g fill="rgb(255,204,0)" text-rendering="geometricPrecision" shape-rendering="geometricPrecision" transform="matrix(1,0,0,1,20,-186)" stroke="rgb(255,204,0)">
      <rect x="200.5" width="520" height="30" y="414" clip-path="url(#clipPath2)" stroke="none"/>
    </g>
    <g text-rendering="geometricPrecision" stroke-miterlimit="1.45" shape-rendering="geometricPrecision" transform="matrix(1,0,0,1,20,-186)" stroke-linecap="butt">
      <rect fill="none" x="200.5" width="520" height="30" y="414" clip-path="url(#clipPath2)"/>
      <text x="427.2686" xml:space="preserve" y="433.1543" clip-path="url(#clipPath2)" font-family="sans-serif" stroke="none">executable</text>
    </g>
    <g fill="rgb(153,204,255)" text-rendering="geometricPrecision" shape-rendering="geometricPrecision" transform="matrix(1,0,0,1,20,-186)" stroke="rgb(153,204,255)">
      <rect x="200.5" width="211" height="30" y="201" clip-path="url(#clipPath2)" stroke="none"/>
    </g>
    <g text-rendering="geometricPrecision" stroke-miterlimit="1.45" shape-rendering="geometricPrecision" transform="matrix(1,0,0,1,20,-186)" stroke-linecap="butt">
      <rect fill="none" x="200.5" width="211" height="30" y="201" clip-path="url(#clipPath2)"/>
      <text x="284.3998" xml:space="preserve" y="220.656" clip-path="url(#clipPath2)" font-family="'Noto Sans Mono'" stroke="none">main.o</text>
    </g>
    <g fill="rgb(153,204,255)" text-rendering="geometricPrecision" shape-rendering="geometricPrecision" transform="matrix(1,0,0,1,20,-186)" stroke="rgb(153,204,255)">
      <rect x="-5" width="158" height="30" y="201" clip-path="url(#clipPath2)" stroke="none"/>
    </g>
    <g text-rendering="geometricPrecision" stroke-miterlimit="1.45" shape-rendering="geometricPrecision" transform="matrix(1,0,0,1,20,-186)" stroke-linecap="butt">
      <rect fill="none" x="-5" width="158" height="30" y="201" clip-path="url(#clipPath2)"/>
      <text x="55.9999" xml:space="preserve" y="220.656" clip-path="url(#clipPath2)" font-family="'Noto Sans Mono'" stroke="none">foo.o</text>
    </g>
    <g fill="rgb(204,255,255)" text-rendering="geometricPrecision" shape-rendering="geometricPrecision" transform="matrix(1,0,0,1,20,-186)" stroke="rgb(204,255,255)">
      <rect x="411.5" width="309" height="183" y="231" clip-path="url(#clipPath2)" stroke="none"/>
    </g>
    <g text-rendering="geometricPrecision" stroke-miterlimit="1.45" shape-rendering="geometricPrecision" transform="matrix(1,0,0,1,20,-186)" stroke-linecap="butt">
      <rect fill="none" x="411.5" width="309" height="183" y="231" clip-path="url(#clipPath2)"/>
      <text x="447.199" xml:space="preserve" y="302.6399" clip-path="url(#clipPath2)" font-family="'Noto Sans Mono'" stroke="none">void   mulle_atinit( void (*f)())</text>
      <text x="447.199" xml:space="preserve" y="318.984" clip-path="url(#clipPath2)" font-family="'Noto Sans Mono'" stroke="none">{</text>
      <text x="447.199" xml:space="preserve" y="335.3281" clip-path="url(#clipPath2)" font-family="'Noto Sans Mono'" stroke="none">   ...</text>
      <text x="447.199" xml:space="preserve" y="351.6722" clip-path="url(#clipPath2)" font-family="'Noto Sans Mono'" stroke="none">}</text>
    </g>
    <g fill="rgb(153,204,255)" text-rendering="geometricPrecision" shape-rendering="geometricPrecision" transform="matrix(1,0,0,1,20,-186)" stroke="rgb(153,204,255)">
      <rect x="411.5" width="309" height="30" y="201" clip-path="url(#clipPath2)" stroke="none"/>
    </g>
    <g text-rendering="geometricPrecision" stroke-miterlimit="1.45" shape-rendering="geometricPrecision" transform="matrix(1,0,0,1,20,-186)" stroke-linecap="butt">
      <rect fill="none" x="411.5" width="309" height="30" y="201" clip-path="url(#clipPath2)"/>
      <text x="504.7995" xml:space="preserve" y="220.656" clip-path="url(#clipPath2)" font-family="'Noto Sans Mono'" stroke="none">libmulle-atinit.a</text>
    </g>
  </g>
</svg>
