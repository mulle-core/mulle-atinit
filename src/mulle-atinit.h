#include <stdint.h>

/*
 *  (c) 2019 nat ORGANIZATION
 *
 *  version:  major, minor, patch
 */
#define MULLE_ATINIT_VERSION  ((0 << 20) | (0 << 8) | 1)


static inline unsigned int   mulle_atinit_get_version_major( void)
{
   return( MULLE_ATINIT_VERSION >> 20);
}


static inline unsigned int   mulle_atinit_get_version_minor( void)
{
   return( (MULLE_ATINIT_VERSION >> 8) & 0xFFF);
}


static inline unsigned int   mulle_atinit_get_version_patch( void)
{
   return( MULLE_ATINIT_VERSION & 0xFF);
}


extern uint32_t   mulle_atinit_get_version( void);


#ifdef __linux


#include "include.h"

#include <stdlib.h>
#include <stdio.h>


/*
   Add other library headers here like so, for exposure to library
   consumers.

   # include "foo.h"
*/


//
// this is called by a `__attribute__((constructor))`
//
// __attribute__((constructor))
// static void load( void)
// {
//    mulle_atinit( f, NULL, priority);
// }
//
extern void   _mulle_atinit( void (*f)( void *), void *userinfo, int priority);

static inline void   mulle_atinit_fail( void (*f)( void *), void *userinfo)
{
   char   *s;

   //
   // If not available, warn and execute anyway and hope for the best.
   // Many mulle-testallocator tests are in C and they don't care.
   //
   s = getenv( "MULLE_ATINIT_FAILURE");
   if( ! s || *s > '0')
      fprintf( stderr, "_mulle_atinit is not available yet, bummer\n");
   if( s && *s > '1')
      abort();
   (*f)( userinfo);
}


static inline void   mulle_atinit( void (*f)( void *), void *userinfo, int priority)
{
   void  (*p_mulle_atinit)( void (*f)( void *), void *, int);
   extern void   *dlsym( void *, const char *);
   char          *s;

   // don't want to force _GNU_SOURCE here so we use 0
   p_mulle_atinit = dlsym( 0, "_mulle_atinit");
   if( ! p_mulle_atinit)
   {
      mulle_atinit_fail( f, userinfo);
      return;
   }
   (*p_mulle_atinit)( f, userinfo, priority);
}


#else

static inline void   mulle_atinit( void (*f)( void *), void *userinfo, int priority)
{
   (*f)( userinfo);
}

#endif
