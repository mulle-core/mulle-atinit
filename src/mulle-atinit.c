#define _GNU_SOURCE

#include "include-private.h"

#include "mulle-atinit.h"


int   __MULLE_ATINIT_ranlib__;


uint32_t   mulle_atinit_get_version( void)
{
   return( MULLE_ATINIT_VERSION);
}


#ifdef __linux

#include <string.h>

#include "mulle-atinit.inc"

#endif

