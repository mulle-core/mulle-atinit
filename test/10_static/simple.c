#define _GNU_SOURCE

#include <mulle-atinit/mulle-atinit.h>
#include <stdio.h>


static void   a( void *s)
{
   printf( "%s: \"%s\"\n", __FUNCTION__, s);
}


static void   b( void *s)
{
   printf( "%s: \"%s\"\n", __FUNCTION__, s);
}


static void   c( void *s)
{
   printf( "%s: \"%s\"\n", __FUNCTION__, s);
}



int  main( void)
{
   mulle_atinit( a, "first", 0);
   mulle_atinit( b, "mid", 0);
   mulle_atinit( c, "last", 0);

   return( 0);
}