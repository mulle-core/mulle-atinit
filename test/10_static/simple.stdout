a: "first"
b: "mid"
c: "last"
